diff --git a/drivers/pmdomain/sunxi/sun55i-pck600.c b/drivers/pmdomain/sunxi/sun55i-pck600.c
index c7ab51514531..79324404decd 100644
--- a/drivers/pmdomain/sunxi/sun55i-pck600.c
+++ b/drivers/pmdomain/sunxi/sun55i-pck600.c
@@ -104,6 +104,10 @@ static int sunxi_pck600_power_on(struct generic_pm_domain *domain)
 static int sunxi_pck600_power_off(struct generic_pm_domain *domain)
 {
 	struct sunxi_pck600_pd *pd = to_sunxi_pd(domain);
+	if (!strcmp(pd->genpd.name, "GPU_TOP") || !strcmp(pd->genpd.name, "GPU_CORE")) {
+		// Skip GPU power off
+		return 0;
+	}
 
 	return sunxi_pck600_pd_set_power(pd, false);
 }
