diff --git a/drivers/pmdomain/sunxi/sun55i-pck600.c b/drivers/pmdomain/sunxi/sun55i-pck600.c
index c7ab51514531..a94aa3b39301 100644
--- a/drivers/pmdomain/sunxi/sun55i-pck600.c
+++ b/drivers/pmdomain/sunxi/sun55i-pck600.c
@@ -163,7 +163,7 @@ static int sunxi_pck600_probe(struct platform_device *pdev)
 		struct sunxi_pck600_pd *pd = &pck->pds[i];
 
 		pd->genpd.name = desc->pd_names[i];
-		pd->genpd.power_off = sunxi_pck600_power_off;
+		// pd->genpd.power_off = sunxi_pck600_power_off;
 		pd->genpd.power_on = sunxi_pck600_power_on;
 		pd->base = base + PPU_REG_SIZE * i;
 
