From 9f15c53e9df0c7ff778b306bd0ec6619778a4a60 Mon Sep 17 00:00:00 2001
From: Nick Alilovic <nickalilovic@gmail.com>
Date: Thu, 30 Apr 2026 09:34:47 -0400
Subject: [PATCH 1/2] Add Radxa A7A DTS changes

---
 .../dts/allwinner/sun60i-a733-cubie-a7a.dts   | 268 ++++--------------
 1 file changed, 60 insertions(+), 208 deletions(-)

diff --git a/arch/arm64/boot/dts/allwinner/sun60i-a733-cubie-a7a.dts b/arch/arm64/boot/dts/allwinner/sun60i-a733-cubie-a7a.dts
index cc6dece5cd47..b31c1b252528 100644
--- a/arch/arm64/boot/dts/allwinner/sun60i-a733-cubie-a7a.dts
+++ b/arch/arm64/boot/dts/allwinner/sun60i-a733-cubie-a7a.dts
@@ -6,7 +6,7 @@
 #define TYPEC_DP 1
 /{
 	board = "A733", "A733-CUBIE-A7A-AXP318";
-	compatible = "radxa,cubie-a7a", "arm,sun60iw2p1";
+	compatible = "radxa,cubie-a7a", "arm,sun60iw2p1", "allwinner,sun60i-a733";
 
 	reg_usb0_vbus: usb0-vbus {
 		compatible = "regulator-fixed";
@@ -52,18 +52,6 @@ wifi_chip_en: wifi_chip_en {
 		regulator-boot-on;
 	};
 
-	eeprom_wp: eeprom-wp {
-		compatible = "regulator-fixed";
-		regulator-name = "eeprom_wp";
-		regulator-min-microvolt = <3300000>;
-		regulator-max-microvolt = <3300000>;
-		regulator-enable-ramp-delay = <1000>;
-		gpio = <&pio PC 7 GPIO_ACTIVE_HIGH>;
-		enable-active-high;
-		regulator-always-on;
-		regulator-boot-on;
-	};
-
 	fan0: pwm-fan {
 		compatible = "pwm-fan";
 		#cooling-cells = <2>;
@@ -798,7 +786,7 @@ &ufs {
 	vcc-supply = <&reg_dldo6>;
 	vccq-supply = <&reg_dcdc8>;
 	vccq2-supply = <&reg_dcdc8>;
-	status = "okay";
+	status = "disabled";
 };
 
 &sdc1 {
@@ -970,21 +958,6 @@ &pcie_rc {
 	status = "okay";
 };
 
-&i2c0 {
-	clock-frequency = <400000>;
-	pinctrl-0 = <&twi0_pins_default>;
-	pinctrl-1 = <&twi0_pins_sleep>;
-	pinctrl-names = "default", "sleep";
-	status = "okay";
-
-		eeprom@50 {
-			compatible = "atmel,24c16";
-			reg = <0x50>;
-			//num-addresses = <1>;
-			status = "okay";
-		};
-	};
-
 &r_i2c0 {
 	clock-frequency = <400000>;
 	status = "okay";
@@ -1109,22 +1082,22 @@ reg_aldo1: aldo1 {
 				regulator-always-on;
 			};
 			reg_aldo2: aldo2 {
-				regulator-name = "axp8191-aldo2";
-				regulator-min-microvolt = <500000>;
-				regulator-max-microvolt = <3400000>;
-				regulator-enable-ramp-delay = <1000>;
+				//regulator-name = "axp8191-aldo2";
+				//regulator-min-microvolt = <500000>;
+				//regulator-max-microvolt = <3400000>;
+				//regulator-enable-ramp-delay = <1000>;
 			};
 			reg_aldo3: aldo3 {
-				regulator-name = "axp8191-aldo3";
-				regulator-min-microvolt = <500000>;
-				regulator-max-microvolt = <3400000>;
-				regulator-enable-ramp-delay = <1000>;
+				//regulator-name = "axp8191-aldo3";
+				//regulator-min-microvolt = <500000>;
+				//regulator-max-microvolt = <3400000>;
+				//regulator-enable-ramp-delay = <1000>;
 			};
 			reg_aldo4: aldo4 {
-				regulator-name = "axp8191-aldo4";
-				regulator-min-microvolt = <500000>;
-				regulator-max-microvolt = <3400000>;
-				regulator-enable-ramp-delay = <1000>;
+				//regulator-name = "axp8191-aldo4";
+				//regulator-min-microvolt = <500000>;
+				//regulator-max-microvolt = <3400000>;
+				//regulator-enable-ramp-delay = <1000>;
 			};
 			reg_aldo5: aldo5 {
 				regulator-name = "axp8191-aldo5";
@@ -1133,10 +1106,10 @@ reg_aldo5: aldo5 {
 				regulator-enable-ramp-delay = <1000>;
 			};
 			reg_aldo6: aldo6 {
-				regulator-name = "axp8191-aldo6";
-				regulator-min-microvolt = <500000>;
-				regulator-max-microvolt = <3400000>;
-				regulator-enable-ramp-delay = <1000>;
+				//regulator-name = "axp8191-aldo6";
+				//regulator-min-microvolt = <500000>;
+				//regulator-max-microvolt = <3400000>;
+				//regulator-enable-ramp-delay = <1000>;
 			};
 			reg_bldo1: bldo1 {
 				regulator-name = "axp8191-bldo1";
@@ -1151,10 +1124,10 @@ reg_bldo2: bldo2 {
 				regulator-enable-ramp-delay = <1000>;
 			};
 			reg_bldo3: bldo3 {
-				regulator-name = "axp8191-bldo3";
-				regulator-min-microvolt = <500000>;
-				regulator-max-microvolt = <3400000>;
-				regulator-enable-ramp-delay = <1000>;
+				//regulator-name = "axp8191-bldo3";
+				//regulator-min-microvolt = <500000>;
+				//regulator-max-microvolt = <3400000>;
+				//regulator-enable-ramp-delay = <1000>;
 			};
 			reg_bldo4: bldo4 {
 				regulator-name = "axp8191-bldo4";
@@ -1191,10 +1164,10 @@ reg_cldo3: cldo3 {
 				regulator-always-on;
 			};
 			reg_cldo4: cldo4 {
-				regulator-name = "axp8191-cldo4";
-				regulator-min-microvolt = <500000>;
-				regulator-max-microvolt = <3500000>;
-				regulator-enable-ramp-delay = <1000>;
+				//regulator-name = "axp8191-cldo4";
+				//regulator-min-microvolt = <500000>;
+				//regulator-max-microvolt = <3500000>;
+				//regulator-enable-ramp-delay = <1000>;
 			};
 			reg_cldo5: cldo5 {
 				regulator-name = "axp8191-cldo5";
@@ -1213,28 +1186,28 @@ reg_dldo1: dldo1 {
 				regulator-always-on;
 			};
 			reg_dldo2: dldo2 {
-				regulator-name = "axp8191-dldo2";
-				regulator-min-microvolt = <500000>;
-				regulator-max-microvolt = <3400000>;
-				regulator-enable-ramp-delay = <1000>;
+				//regulator-name = "axp8191-dldo2";
+				//regulator-min-microvolt = <500000>;
+				//regulator-max-microvolt = <3400000>;
+				//regulator-enable-ramp-delay = <1000>;
 			};
 			reg_dldo3: dldo3 {
-				regulator-name = "axp8191-dldo3";
-				regulator-min-microvolt = <500000>;
-				regulator-max-microvolt = <3400000>;
-				regulator-enable-ramp-delay = <1000>;
+				//regulator-name = "axp8191-dldo3";
+				//regulator-min-microvolt = <500000>;
+				//regulator-max-microvolt = <3400000>;
+				//regulator-enable-ramp-delay = <1000>;
 			};
 			reg_dldo4: dldo4 {
-				regulator-name = "axp8191-dldo4";
-				regulator-min-microvolt = <500000>;
-				regulator-max-microvolt = <3400000>;
-				regulator-enable-ramp-delay = <1000>;
+				//regulator-name = "axp8191-dldo4";
+				//regulator-min-microvolt = <500000>;
+				//regulator-max-microvolt = <3400000>;
+				//regulator-enable-ramp-delay = <1000>;
 			};
 			reg_dldo5: dldo5 {
-				regulator-name = "axp8191-dldo5";
-				regulator-min-microvolt = <500000>;
-				regulator-max-microvolt = <3400000>;
-				regulator-enable-ramp-delay = <1000>;
+				//regulator-name = "axp8191-dldo5";
+				//regulator-min-microvolt = <500000>;
+				//regulator-max-microvolt = <3400000>;
+				//regulator-enable-ramp-delay = <1000>;
 			};
 			reg_dldo6: dldo6 {
 				regulator-name = "axp8191-dldo6";
@@ -1253,30 +1226,30 @@ reg_eldo1: eldo1 {
 				regulator-always-on;
 			};
 			reg_eldo2: eldo2 {
-				regulator-name = "axp8191-eldo2";
-				regulator-min-microvolt = <500000>;
-				regulator-max-microvolt = <3400000>;
-				regulator-enable-ramp-delay = <1000>;
-				regulator-boot-on;
-				regulator-always-on;
+				//regulator-name = "axp8191-eldo2";
+				//regulator-min-microvolt = <500000>;
+				//regulator-max-microvolt = <3400000>;
+				//regulator-enable-ramp-delay = <1000>;
+				//regulator-boot-on;
+				//regulator-always-on;
 			};
 			reg_eldo3: eldo3 {
-				regulator-name = "axp8191-eldo3";
-				regulator-min-microvolt = <500000>;
-				regulator-max-microvolt = <3400000>;
-				regulator-enable-ramp-delay = <1000>;
+				//regulator-name = "axp8191-eldo3";
+				//regulator-min-microvolt = <500000>;
+				//regulator-max-microvolt = <3400000>;
+				//regulator-enable-ramp-delay = <1000>;
 			};
 			reg_eldo4: eldo4 {
-				regulator-name = "axp8191-eldo4";
-				regulator-min-microvolt = <500000>;
-				regulator-max-microvolt = <3400000>;
-				regulator-enable-ramp-delay = <1000>;
+				//regulator-name = "axp8191-eldo4";
+				//regulator-min-microvolt = <500000>;
+				//regulator-max-microvolt = <3400000>;
+				//regulator-enable-ramp-delay = <1000>;
 			};
 			reg_eldo5: eldo5 {
-				regulator-name = "axp8191-eldo5";
-				regulator-min-microvolt = <500000>;
-				regulator-max-microvolt = <3400000>;
-				regulator-enable-ramp-delay = <1000>;
+				//regulator-name = "axp8191-eldo5";
+				//regulator-min-microvolt = <500000>;
+				//regulator-max-microvolt = <3400000>;
+				//regulator-enable-ramp-delay = <1000>;
 			};
 			reg_eldo6: eldo6 {
 				regulator-name = "axp8191-eldo6";
@@ -1297,127 +1270,6 @@ reg_dc1sw2: dc1sw2 {
 				regulator-always-on;
 			};
 		};
-
-		virtual-dcdc1 {
-			compatible = "xpower-vregulator,dcdc1";
-			dcdc1-supply = <&reg_dcdc1>;
-		};
-		virtual-dcdc2 {
-			compatible = "xpower-vregulator,dcdc2";
-			dcdc2-supply = <&reg_dcdc2>;
-		};
-		virtual-dcdc3 {
-			compatible = "xpower-vregulator,dcdc3";
-			dcdc3-supply = <&reg_dcdc3>;
-		};
-		virtual-dcdc4 {
-			compatible = "xpower-vregulator,dcdc4";
-			dcdc4-supply = <&reg_dcdc4>;
-		};
-		virtual-dcdc5 {
-			compatible = "xpower-vregulator,dcdc5";
-			dcdc5-supply = <&reg_dcdc5>;
-		};
-		virtual-dcdc6 {
-			compatible = "xpower-vregulator,dcdc6";
-			dcdc6-supply = <&reg_dcdc6>;
-		};
-		virtual-dcdc7 {
-			compatible = "xpower-vregulator,dcdc7";
-			dcdc7-supply = <&reg_dcdc7>;
-		};
-		virtual-dcdc8 {
-			compatible = "xpower-vregulator,dcdc8";
-			dcdc8-supply = <&reg_dcdc8>;
-		};
-		virtual-dcdc9 {
-			compatible = "xpower-vregulator,dcdc9";
-			dcdc9-supply = <&reg_dcdc9>;
-		};
-		virtual-rtcldo {
-			compatible = "xpower-vregulator,rtcldo";
-			rtcldo-supply = <&reg_rtcldo>;
-		};
-		virtual-aldo1 {
-			compatible = "xpower-vregulator,aldo1";
-			aldo1-supply = <&reg_aldo1>;
-		};
-		virtual-aldo2 {
-			compatible = "xpower-vregulator,aldo2";
-			aldo2-supply = <&reg_aldo2>;
-		};
-		virtual-aldo3 {
-			compatible = "xpower-vregulator,aldo3";
-			aldo3-supply = <&reg_aldo3>;
-		};
-		virtual-aldo4 {
-			compatible = "xpower-vregulator,aldo4";
-			aldo4-supply = <&reg_aldo4>;
-		};
-		virtual-aldo5 {
-			compatible = "xpower-vregulator,aldo5";
-			aldo5-supply = <&reg_aldo5>;
-		};
-		virtual-aldo6 {
-			compatible = "xpower-vregulator,aldo6";
-			aldo6-supply = <&reg_aldo6>;
-		};
-		virtual-dldo1 {
-			compatible = "xpower-vregulator,dldo1";
-			dldo1-supply = <&reg_dldo1>;
-		};
-		virtual-dldo2 {
-			compatible = "xpower-vregulator,dldo2";
-			dldo2-supply = <&reg_dldo2>;
-		};
-		virtual-dldo3 {
-			compatible = "xpower-vregulator,dldo3";
-			dldo3-supply = <&reg_dldo3>;
-		};
-		virtual-dldo4 {
-			compatible = "xpower-vregulator,dldo4";
-			dldo4-supply = <&reg_dldo4>;
-		};
-		virtual-dldo5 {
-			compatible = "xpower-vregulator,dldo5";
-			dldo5-supply = <&reg_dldo5>;
-		};
-		virtual-dldo6 {
-			compatible = "xpower-vregulator,dldo6";
-			dldo6-supply = <&reg_dldo6>;
-		};
-		virtual-eldo1 {
-			compatible = "xpower-vregulator,eldo1";
-			eldo1-supply = <&reg_eldo1>;
-		};
-		virtual-eldo2 {
-			compatible = "xpower-vregulator,eldo2";
-			eldo2-supply = <&reg_eldo2>;
-		};
-		virtual-eldo3 {
-			compatible = "xpower-vregulator,eldo3";
-			eldo3-supply = <&reg_eldo3>;
-		};
-		virtual-eldo4 {
-			compatible = "xpower-vregulator,eldo4";
-			eldo4-supply = <&reg_eldo4>;
-		};
-		virtual-eldo5 {
-			compatible = "xpower-vregulator,eldo5";
-			eldo5-supply = <&reg_eldo5>;
-		};
-		virtual-eldo6 {
-			compatible = "xpower-vregulator,eldo6";
-			eldo6-supply = <&reg_eldo6>;
-		};
-		virtual-dc1sw1 {
-			compatible = "xpower-vregulator,dc1sw1";
-			dc1sw-supply = <&reg_dc1sw1>;
-		};
-		virtual-dc1sw2 {
-			compatible = "xpower-vregulator,dc1sw2";
-			dc1sw-supply = <&reg_dc1sw2>;
-		};
 	};
 };
 /*Disable soc rtc*/
-- 
2.51.0

